src/Workspaces/Core/Portable/Utilities/DocumentTextChangeMerger.cs

```diff
@@ -0,0 +1,241 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Utilities
+{
+    internal class DocumentTextChangeMerger : IIntervalIntrospector<TextChange>
```

Comment:
this type is effectively just extracted from the batch fixer.  if you look commit by commit you will see how it is virtually identical to that code.